homeassistant/components/fritz/sensor.py

```diff
@@ -72,6 +72,36 @@ def _retrieve_gb_received_state(status: FritzStatus, last_value: str) -> float:
     return round(status.bytes_received * 8 / 1024 / 1024 / 1024, 1)  # type: ignore[no-any-return]
 
 
+def _retrieve_link_kb_s_sent_state(status: FritzStatus, last_value: str) -> float:
+    """Return upload link rate."""
+    return round(status.max_linked_bit_rate[0] / 1024, 1)  # type: ignore[no-any-return]
+
+
+def _retrieve_link_kb_s_received_state(status: FritzStatus, last_value: str) -> float:
+    """Return download link rate."""
+    return round(status.max_linked_bit_rate[1] / 1024, 1)  # type: ignore[no-any-return]
+
+
+def _retrieve_link_noise_margin_sent_state(status: FritzStatus, last_value: str) -> float:
+    """Return upload noise margin."""
+    return status.noise_margin[0]  # type: ignore[no-any-return]
+
+
+def _retrieve_link_noise_margin_received_state(status: FritzStatus, last_value: str) -> float:
+    """Return download noise margin."""
+    return status.noise_margin[1]  # type: ignore[no-any-return]
+
+
+def _retrieve_link_attenuation_sent_state(status: FritzStatus, last_value: str) -> float:
+    """Return upload line attenuation."""
+    return status.attenuation[0]  # type: ignore[no-any-return]
+
+
+def _retrieve_link_attenuation_received_state(status: FritzStatus, last_value: str) -> float:
+    """Return download line attenuation."""
+    return status.attenuation[1]  # type: ignore[no-any-return]
+
+
```

Comment:
What happens on lines without those values ? Like for example FTTC/FTTH ?

Reply:
For non DSL Lines where these values are not available it fails gracefully. Tested with ethernet WAN connection on Fritz!Box4040. 
The error is caught in the upstream library. The output is either null or 0 (have to check at the Fritz!box4040 test setup at the weekend...

Reply:
I think we should create the sensor only if the device return those values, and not always

Reply:
That seems a good solution, My skill in Smarthome programming is not good enough to do conditional sensor values. Is a simple if statement enough?

Reply:
I this we should add a call to:

```python
dsl: bool = fritzbox_tools.connection.call_action("WANDSLInterfaceConfig:1", "GetInfo")["NewEnable"]
```

and base `entities.append` on that.

Simone

Reply:
Are you planning to proceed coding this part ?

Simone

Reply:
Code updated ;-)